libafl_qemu/src/qemu/qemu_config.rs

```diff
@@ -0,0 +1,401 @@
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "-machine {}", self.name)
+    }
+}
+
+impl<R: AsRef<str>> From<R> for Machine {
+    fn from(name: R) -> Self {
+        Self {
+            name: name.as_ref().to_string(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, strum_macros::Display)]
+pub enum Snapshot {
+    #[strum(serialize = "-snapshot")]
+    ENABLE,
+    #[strum(serialize = "")]
+    DISABLE,
+}
+
+impl From<bool> for Snapshot {
+    fn from(snapshot: bool) -> Self {
+        if snapshot {
+            Snapshot::ENABLE
+        } else {
+            Snapshot::DISABLE
+        }
+    }
+}
+
+/// When set to DISABLE, corresponds to the `-S` option of QEMU.
+#[derive(Debug, Clone, strum_macros::Display)]
+pub enum StartCPU {
+    #[strum(serialize = "")]
+    ENABLE,
+    #[strum(serialize = "-S")]
+    DISABLE,
+}
+
+impl From<bool> for StartCPU {
+    fn from(start_cpu: bool) -> Self {
+        if start_cpu {
+            StartCPU::ENABLE
+        } else {
+            StartCPU::DISABLE
+        }
+    }
+}
+
+#[derive(Debug, Clone, strum_macros::Display)]
+pub enum NoGraphic {
+    #[strum(serialize = "-nographic")]
+    ENABLE,
+    #[strum(serialize = "")]
+    DISABLE,
+}
+
+impl From<bool> for NoGraphic {
+    fn from(no_graphic: bool) -> Self {
+        if no_graphic {
+            NoGraphic::ENABLE
+        } else {
+            NoGraphic::DISABLE
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum RamSize {
+    MB(u32),
+    GB(u32),
+}
+
+impl Display for RamSize {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            RamSize::MB(mb) => write!(f, "-m {mb}M"),
+            RamSize::GB(gb) => write!(f, "-m {gb}G"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SmpCpus {
+    pub cpus: u32,
+}
+
+impl Display for SmpCpus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "-smp {}", self.cpus)
+    }
+}
+
+#[derive(Debug, Clone, strum_macros::Display)]
+pub enum VgaPci {
+    #[strum(serialize = "-device VGA")]
+    ENABLE,
+    #[strum(serialize = "")]
+    DISABLE,
+}
+
+impl From<bool> for VgaPci {
+    fn from(vga_pci: bool) -> Self {
+        if vga_pci {
+            VgaPci::ENABLE
+        } else {
+            VgaPci::DISABLE
+        }
+    }
+}
+
+#[cfg(emulation_mode = "usermode")]
+#[derive(Debug, Clone)]
+pub struct Program {
+    path: PathBuf,
+}
+
+#[cfg(emulation_mode = "usermode")]
+impl Display for Program {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.path.to_str().unwrap())
+    }
+}
+
+#[cfg(emulation_mode = "usermode")]
+impl<R: AsRef<Path>> From<R> for Program {
+    fn from(path: R) -> Self {
+        Self {
+            path: path.as_ref().to_path_buf(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, libafl_derive::Display, TypedBuilder, Getters)]
+#[builder(build_method(into = Result<Qemu, QemuInitError>), builder_method(vis = "pub(crate)",
```

Comment:
Why only `pub(crate)` btw?

Reply:
The `QemuConfig::builder` is called in `Qemu::builder()` which is the only place where it should be called, in this way we can enforce the 1 to 1 matching of `Qemu` and `QemuConfig`

lmk if this make sense to you and clarifies :)

Reply:
Ah right makes sense